src/headers/frame_header.rs

```diff
@@ -0,0 +1,494 @@
+    #[all_default]
+    #[default(true)]
+    all_default: bool,
+
+    #[default(FrameType::RegularFrame)]
+    frame_type: FrameType,
+
+    #[coder(Bits(1))]
+    #[default(Encoding::VarDCT)]
+    encoding: Encoding,
+
+    // TODO(Tomáš Král): nicer abstraction for bitflags
+    #[default(0)]
+    flags: u64,
+
+    #[default(false)]
+    #[condition(!nonserialized.xyb_encoded)]
+    do_ycbcr: bool,
+
+    #[coder(Bits(2))]
+    #[default([0, 0, 0])]
+    #[condition(do_ycbcr && flags & 0x20 == 0)]
+    jpeg_upsampling: [u32; 3],
+
+    #[coder(u2S(1, 2, 4, 8))]
+    #[default(1)]
+    #[condition(flags & 0x20 == 0)]
+    upsampling: u32,
+
+    #[size_coder(explicit(nonserialized.num_extra_channels))]
+    #[coder(u2S(1, 2, 4, 8))]
+    #[default(vec![1; nonserialized.num_extra_channels as usize])]
+    #[condition(flags & 0x20 == 0)]
+    ec_upsampling: Vec<u32>,
+
+    #[coder(Bits(2))]
+    #[default(1)]
+    #[condition(encoding == Encoding::Modular)]
+    group_size_shift: u32,
+
+    #[coder(Bits(3))]
+    #[default(3)]
+    #[condition(encoding == Encoding::VarDCT && nonserialized.xyb_encoded)]
+    x_qm_scale: u32,
+
+    #[coder(Bits(3))]
+    #[default(2)]
+    #[condition(encoding == Encoding::VarDCT && nonserialized.xyb_encoded)]
+    b_qm_scale: u32,
+
+    #[condition(frame_type != FrameType::ReferenceOnly)]
+    #[default(Passes::default())]
+    passes: Passes,
+
+    #[coder(u2S(1, 2, 3, 4))]
+    #[default(0)]
+    #[condition(frame_type == FrameType::LFFrame)]
+    lf_level: u32,
+
+    #[default(false)]
+    #[condition(frame_type != FrameType::LFFrame)]
+    have_crop: bool,
+
+    #[coder(u2S(Bits(8), Bits(11) + 256, Bits(14) + 2304, Bits(30) + 18688))]
+    #[default(0)]
+    #[condition(have_crop && frame_type != FrameType::ReferenceOnly)]
+    x0: u32,
+
+    #[coder(u2S(Bits(8), Bits(11) + 256, Bits(14) + 2304, Bits(30) + 18688))]
+    #[default(0)]
+    #[condition(have_crop && frame_type != FrameType::ReferenceOnly)]
+    y0: u32,
+
+    #[coder(u2S(Bits(8), Bits(11) + 256, Bits(14) + 2304, Bits(30) + 18688))]
+    #[default(0)]
+    #[condition(have_crop)]
+    width: u32,
+
+    #[coder(u2S(Bits(8), Bits(11) + 256, Bits(14) + 2304, Bits(30) + 18688))]
+    #[default(0)]
+    #[condition(have_crop)]
+    height: u32,
+
+    /* "normal_frame" denotes the condition !all_default
+    && (frame_type == kRegularFrame || frame_type == kSkipProgressive) */
+    #[default(BlendingInfo::default())]
+    #[condition(frame_type == FrameType::RegularFrame || frame_type == FrameType::SkipProgressive)]
+    #[nonserialized(num_extra_channels : nonserialized.num_extra_channels,
+        have_crop : have_crop, x0: x0, y0: y0, width: width, height: height,
+        img_width: nonserialized.img_width, img_height: nonserialized.img_height)]
+    blending_info: BlendingInfo,
+
+    #[size_coder(explicit(nonserialized.num_extra_channels))]
+    #[default(vec![BlendingInfo::default(); nonserialized.num_extra_channels as usize])]
```

Comment:
I don't know how complicated it would be to do so, but it might be nicer if `#[default(...)]` would define the default of *vector elements*, instead of the whole vector. This would also help to avoid some code repetition. WDYT?

Reply:
It wouldn't be possible with`#[default] because that needs to be of the same type as the field for the `Default` trait. So I added `#[default_element]` attribute. But I still feel like the solution is a bit dirty...